Improvement: Remove hardcoded version in Salt formulas

gdemonet opened the pull request:
Here is a simplistic solution to the multiple versions problem: we wanted to use multiple Salt environments for having multiple product versions living next to one another. To know which version is being used then, we can simply rely on the `{{ saltenv }}` Jinja variable.

Along the way, we realized another hardcoded info, i.e. *where we mount the ISO*, could not be inferred completely from the active `saltenv`, but instead was a fully customizable value without any implications for the product well-being. We use the newly introduced configuration file to add such info.

Fixes: GH-651

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @ballot-scality

* @thomasdanan

* @cmonfort

* @alxf

* @monpote

* @nicolast

* @slaperche-scality

* @zempashi

* @teddyandrieux

* @carlito-scality

* @tcarmet

* @alexandre-allard-scality

* @ramanduh

* @sayf-eddine-scality

* @nootal

* @lucieleonard

* @gdemonet

* @jbertran

gdemonet committed:
Use saltenv for product version info

Issue: GH-651

gdemonet committed:
Add isoMountpoint configuration option

Note: there is no default value for this option yet, we'd need a better
mechanism for loading the config file with defaults.

Issue: GH-651

gdemonet committed:
Use shorter Make syntax for creating directories

Issue: GH-651

NicolasT commented on Vagrantfile:
```diff
@@ -108,6 +108,7 @@ mkdir -p /etc/metalk8s
 cat > /etc/metalk8s/bootstrap.yaml << EOF
 apiVersion: metalk8s.scality.com/v1alpha1
 kind: BootstrapConfiguration
+isoMountpoint: /srv/scality/metalk8s-$SHORT_VERSION
```
No, we can't have this here. This config is system-wide, across multiple 'versions' of MetalK8s available in the environment.

gdemonet replied:
Right, didn't think this through... So this could go in a pillar provided by us? How would a user edit such pillars if they are located in the read-only mounted ISO?

gdemonet replied:
Moved to a pillar - will need to think about how to customize this...

NicolasT commented on eve/main.yml:
```diff
@@ -221,9 +222,10 @@ stages:
             ssh -F bootstrap_ssh_config bootstrap
             sudo mkdir -p /etc/metalk8s &&
-            scp -F bootstrap_ssh_config bootstrap.yaml bootstrap: &&
+            scp -F bootstrap_ssh_config bootstrap.yaml.in bootstrap: &&
```
How about replacing this with something along the lines of

```
sed "s/@VERSION@/%(prop:metalk8s_short_version)s/g" bootstrap.yaml.in | ssh -F bootstrap_ssh_config bootstrap sudo tee /etc/metalk8s/bootstrap.yaml
```

gdemonet replied:
It's not needed however, since we're not configuring the mountpoint in this file anymore

NicolasT commented on salt/_pillar/metalk8s.py:
```diff
@@ -9,25 +9,46 @@
-    assert 'workloadPlane' in obj['networks']
+
+def _load_extra(config_data):
+    assert 'isoMountpoint' in config_data
```
See before.

gdemonet replied:
Removed

NicolasT commented on salt/metalk8s/kubeadm/init/control-plane/lib.sls:
```diff
@@ -1,7 +1,7 @@
 # Share macro for building image names for each control-plane component
-{% set repository = "localhost:5000/metalk8s-2.0" %}
-{% set kubernetes_version = "1.11.7" %}
+{%- set registry_url = "localhost:5000" -%}
```
This is not really a URL :-P

gdemonet replied:
What would you call it? `registry_host` / `registry_address`?

gdemonet replied:
Used `registry_host`

NicolasT commented on salt/metalk8s/salt/master/files/master_99-metalk8s.conf.j2:
```diff
@@ -1,9 +1,9 @@
 file_roots:
-  metalk8s-{{ metal_version }}:
-    - /srv/scality/metalk8s-2.0/salt
+  {{ saltenv }}:
```
Once we implement multiple versions and upgrades, this will no longer work correctly (but that's not  regression introduced by this PR, just a heads-up)

gdemonet replied:
Should I create a ticket tracking this as debt? I could split the master conf into multiple files, so that declaring file_roots is done in some `/etc/salt/master.d/99-metalk8s-{{ version }}.conf` which won't get overridden by other Salt environments.

gdemonet replied:
Created #800 to track this.

NicolasT reviewed (changes requested):
The ISO path definition in the bootstrap config won't work...

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @ballot-scality

* @thomasdanan

* @cmonfort

* @alxf

* @monpote

* @nicolast

* @slaperche-scality

* @zempashi

* @teddyandrieux

* @carlito-scality

* @tcarmet

* @alexandre-allard-scality

* @ramanduh

* @sayf-eddine-scality

* @nootal

* @lucieleonard

* @gdemonet

* @jbertran





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

NicolasT commented on Vagrantfile:
```diff
@@ -108,6 +108,7 @@ mkdir -p /etc/metalk8s
 cat > /etc/metalk8s/bootstrap.yaml << EOF
 apiVersion: metalk8s.scality.com/v1alpha1
 kind: BootstrapConfiguration
+isoMountpoint: /srv/scality/metalk8s-$SHORT_VERSION
```
Shouldn't this be removed now?

gdemonet replied:
It should, oops

NicolasT commented on eve/main.yml:
```diff
@@ -125,6 +125,7 @@ stages:
             cat > /etc/metalk8s/bootstrap.yaml << END
             apiVersion: metalk8s.scality.com/v1alpha1
             kind: BootstrapConfiguration
+            isoMountpoint: /srv/scality/metalk8s-%(prop:metalk8s_short_version)s
```
Shouldn't this be removed now?

gdemonet replied:
It should, oops

NicolasT reviewed (changes requested):
Some leftovers.

NicolasT commented on Makefile:
```diff
@@ -212,29 +212,34 @@ all-local: $(ALL) ## Build all artifacts in the build tree
 	rm -f $@
 	cp -a $< $@
 
+$(ISO_ROOT)/pillar/metalk8s.sls: pillar/metalk8s.sls.in $(ISO_ROOT)/product.txt
```
Minor: this could be merged with the rule below for `$(ISO_ROOT)/pillar/%.sls`, but given #708 it's not worth it :)

gdemonet replied:
That's what I thought, and we have another `$(ISO_ROOT)/pillar/repositories.sls` which does not follow the same rule, so I was not sure how to write the pattern :/

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @ballot-scality

* @thomasdanan

* @cmonfort

* @alxf

* @monpote

* @nicolast

* @slaperche-scality

* @zempashi

* @teddyandrieux

* @carlito-scality

* @tcarmet

* @alexandre-allard-scality

* @ramanduh

* @sayf-eddine-scality

* @nootal

* @lucieleonard

* @gdemonet

* @jbertran

gdemonet commented:
@bert-e approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @ballot-scality

* @thomasdanan

* @cmonfort

* @alxf

* @monpote

* @nicolast

* @slaperche-scality

* @zempashi

* @teddyandrieux

* @carlito-scality

* @tcarmet

* @alexandre-allard-scality

* @ramanduh

* @sayf-eddine-scality

* @nootal

* @lucieleonard

* @gdemonet

* @jbertran










*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue None.

Goodbye gdemonet.
